libs/pika/async_mpi/src/mpi_future.cpp

```diff
@@ -226,150 +357,95 @@ namespace pika { namespace mpi { namespace experimental {
-                        "flag", debug::dec<>(flag), "status",
-                        debug::hex(status.MPI_ERROR));
-                }
-                else
-                {
-                    auto poll_deb =
-                        mpi_debug.make_timer(1, debug::str<>("Poll - <ERR>"));
-
-                    mpi_debug.error(poll_deb, detail::get_mpi_info(),
-                        debug::str<>("Poll <ERR>"), "MPI_ERROR",
-                        detail::error_message(status.MPI_ERROR), "status",
-                        debug::dec<>(status.MPI_ERROR), "index",
-                        debug::dec<>(index), "flag", debug::dec<>(flag));
-                }
+                std::terminate();
             }
 
-            // No operation completed
-            if (index == MPI_UNDEFINED)
-                break;
-
-            keep_trying = flag;
             if constexpr (mpi_debug.is_enabled())
             {
-                // One operation completed
-                if (keep_trying)
-                {
-                    mpi_debug.debug(debug::str<>("MPI_Testany(set)"),
-                        detail::get_mpi_info(), "request",
-                        debug::hex<8>(requests_vector[std::size_t(index)]),
-                        "vector size", debug::dec<3>(requests_vector.size()),
-                        "non null",
-                        debug::dec<3>(
-                            detail::get_num_active_requests_in_vector()));
-                }
+                mpi_debug.debug(debug::str<>("MPI_Testsome (set)"),
+                    detail::mpi_data_, "request",
+                    debug::hex<8>(detail::mpi_data_.request_vector_[index]));
             }
-            if (result != MPI_SUCCESS)    // Error and operation not completed
-                keep_trying = false;
-            if (keep_trying || result != MPI_SUCCESS)
-            {
-                // Invoke the callback with the status of the completed
-                // operation (status of the request is forwarded to MPI_Testany)
-                request_callback_vector[std::size_t(index)].callback_function(
-                    result);
 
-                // Remove the request from our vector to prevent retesting
-                requests_vector[std::size_t(index)] = MPI_REQUEST_NULL;
+            // decrement before invoking callback to avoid race
+            // if invoked code checks in_flight value
+            size_t inflight = --detail::mpi_data_.in_flight_;
+            --detail::mpi_data_.active_request_vector_size_;
 
-                // Could store only the callbacks, right now the request
-                // is only used for an assert
-                request_callback_vector[std::size_t(index)].request =
-                    MPI_REQUEST_NULL;
+            // Invoke the callback with the status of the completed
+            // operation (status of the request is forwarded to MPI_Testany)
+            detail::mpi_data_.callback_vector_[index](result);
```

Comment:
Maybe?
```suggestion
            detail::mpi_data_.callback_vector_[index](detail::mpi_data_.status_vector_[i].MPI_ERROR);
```
So that the callback can decide what to do with errors. Otherwise there's no point in having it take an MPI status (since `result` is always `MPI_SUCCESS` here).

Reply:
done